crates/proof-of-sql/src/sql/postprocessing/evaluate.rs

```diff
@@ -0,0 +1,252 @@
+use super::{PostprocessingError, PostprocessingResult};
+use crate::base::{
+    database::{
+        try_add_subtract_column_types, try_divide_column_types, try_multiply_column_types,
+        ColumnType, OwnedColumn, OwnedTable,
+    },
+    math::decimal::{scale_scalar, try_into_to_scalar, Precision},
+    scalar::Scalar,
+};
+use core::cmp::Ordering;
+use proof_of_sql_parser::{
+    intermediate_ast::{BinaryOperator, Expression, Literal, UnaryOperator},
+    Identifier,
+};
+
+/// Evaluator that evaluates a `proof_of_sql_parser::intermediate_ast::Expression`
+/// on an `OwnedTable` and returns an `OwnedColumn`.
+pub struct PostprocessingEvaluator<S: Scalar> {
+    table: OwnedTable<S>,
+}
+
+impl<S: Scalar> PostprocessingEvaluator<S> {
+    /// Creates a new `PostprocessingEvaluator` with the given owned table.
+    pub fn new(table: &OwnedTable<S>) -> Self {
+        Self {
+            table: table.clone(),
+        }
+    }
+    /// Evaluates the given expression on the owned table and returns the result.
+    pub fn evaluate(&self, expr: &Expression) -> PostprocessingResult<OwnedColumn<S>> {
+        self.visit_expr(expr)
+    }
+}
+
+// Private interface
+impl<S: Scalar> PostprocessingEvaluator<S> {
+    fn visit_expr(&self, expr: &Expression) -> PostprocessingResult<OwnedColumn<S>> {
```

Comment:
Since `PostprocessingEvaluator` is just a wrapper around `OwnedTable`, why not just make it an extension type? Saves a step when calling this code.
```suggestion
pub trait PostprocessingEvaluator {
    type Scalar: Scalar;
    fn evaluate(&self, expr: &Expression) -> PostprocessingResult<OwnedColumn<Self::Scalar>>;
}
impl<S: Scalar> PostprocessingEvaluator for OwnedTable<S> {
    type Scalar = S;
    fn evaluate(&self, expr: &Expression) -> PostprocessingResult<OwnedColumn<S>> {
        visit_expr(self, expr)
    }
}
fn visit_expr<S: Scalar>(
    table: &OwnedTable<S>,
    expr: &Expression,
) -> PostprocessingResult<OwnedColumn<S>> {
```

Reply:
Good idea. If it is only going to be applied on `OwnedTable` anyway then we may as well put it in `owned_table_evaluation.rs` instead?